activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -998,35 +998,21 @@ def checkout_and_verify(c)
     # about the model. The model needs to pass a specification name to the handler,
     # in order to look up the correct connection pool.
     class ConnectionHandler
-      def self.create_owner_to_pool # :nodoc:
-        Concurrent::Map.new(initial_capacity: 2) do |h, k|
-          # Discard the parent's connection pools immediately; we have no need
-          # of them
-          discard_unowned_pools(h)
-
-          h[k] = Concurrent::Map.new(initial_capacity: 2)
-        end
-      end
-
-      def self.unowned_pool_finalizer(pid_map) # :nodoc:
+      def self.unowned_pool_finalizer(db_configs_map) # :nodoc:
         lambda do |_|
-          discard_unowned_pools(pid_map)
+          db_configs_map.values.each(&:discard_unowned_pool!)
         end
       end
 
-      def self.discard_unowned_pools(pid_map) # :nodoc:
-        pid_map.each do |pid, pools|
-          pools.values.compact.each(&:discard!) unless pid == Process.pid
-        end
-      end
+      FINALIZER = lambda { |_| ActiveSupport::ForkTracker.check! }
```

Comment:
Should we use `ForkTracker` here and in the other places? If someone add a `after_fork { connect_to_redis }` we don't want the object finalizer neither the `disconnect!` to run that, correct?

Reply:
That's a good question. Not sure I have a definitive answer, but I'll braindump a few points:

  - Ideally your server/job-runner/whatever used `Kernel#fork` or `Process#fork` so `ForkTracker.check!` is almost a noop.
  - The only exception to that we have in mind is passenger, we could probably document this, so that people are recommended to all `check!` in Passenger's post fork callback. (or even start talking with passenger people to see if c-level forking is a hard requirement)
  - For now ForkTracker is private API, people are not supposed to register their own callbacks in it.
  - Maybe the actual issue here is the finalizer. I'll try to think about that tomorrow, maybe we can do without it.

Reply:
Looking at the PR that introduced this finalizer (https://github.com/rails/rails/pull/31173), from what I understand the intent here is to discard the connections before the process close them on exit.

So, I'm not to sure where to go from here:

  - Maybe we can make a requirement of calling `check!` if you fork outside of Ruby.
  - We refactor `ForkTracker` and `DatabaseConfig` so that we make sure to only trigger the callbacks we care about?

@rafaelfranca @matthewd opinions welcome.

Reply:
My two thoughts are:
* I think it's fine for the finalizer to call `check!` -- that doesn't _do_ things, it just gives us a chance to do things _if_ a fork has happened and we haven't noticed yet. _If_ someone has managed to do a sneaky fork and this is the first time we notice, it seems acceptable (though a little pointless, because we're _probably_ in the middle of tearing down the process) to make that redis connection. 🤷🏻‍♂️
* it might be undesirable to do a potentially-arbitrary amount of `after_fork` work inside a finalizer, given an assumption that others will use ForkTracker... I'm trying to remember whether that's a Bad Thing To Do, or actually fine?

Reply:
@FooBarWidget can you help us out with some context here, perhaps? :heart:

Given #37312, we're wary that you might be doing OS-level forks... I note the one [mentioned above](https://github.com/rails/rails/pull/37296#discussion_r329018065) then does an exec, so that's not relevant. But do you know whether you have anywhere you could be doing a direct `fork(2)` with a Ruby environment loaded _without_ a matching exec? I have hopes that [this bit](https://github.com/phusion/passenger/blob/b30bdbe66b3f41d108f160e22778549ef82184d3/src/ruby_supportlib/phusion_passenger/preloader_shared_helpers.rb#L108) might be a strong sign the answer's "no", but my lazy search turned up too many incidental matches to be sure.

Reply:
@matthewd Are you asking whether Passenger, or anything else that I'm aware of, performs a directly fork() system call without calling Ruby's Kernel#fork, and at the same time without calling exec()? Then the answer is no: Passenger is definitely not doing that, and I'm not aware of anything else in the Ruby ecosystem that does that.

Reply:
@FooBarWidget yes, that was the question. 

Thanks a lot for the answer, I got confused by some other forks in passenger's codebase.

Reply:
@rafaelfranca @matthewd then I think we can simply drop the safety `check!` calls. The fork hook should be enough.

Reply:
I think we can definitely drop the safety `check!` calls. The way MRI works, it is not safe at all to perform a fork() system call without calling Ruby's Kernel#fork. If you do that you will corrupt MRI state, and your program will crash anyway. So nobody will call a raw fork() system call and continue to run Ruby code in the new process.

Reply:
Perfect, thanks so much @FooBarWidget! 

@casperisfine I do wonder if we need to consider the edge case where an instance becomes unreferenced, then a fork occurs, and then GC runs in the forked process before the `after_fork` is invoked? Given that's theoretically possible even with a short-lived "fork for just a moment" block, and that it can destroy the long-lived parent process's DB connection(s), seems like it might be in the "quite unlikely, but in the lifetime of all AR-using applications, occasionally inevitable" bucket?

Reply:
> an instance becomes unreferenced, then a fork occurs, and then GC runs in the forked process before the `after_fork` is invoked?

I won't say it's totally impossible, but the window would be terribly small.

The after_fork callback is the very first thing we invoke in the child, so the GC would have to trigger between the fork and the `discard_unowned_connections!` call.

> "quite unlikely, but in the lifetime of all AR-using applications, occasionally inevitable" bucket?

I think I'd be more likely to win the lottery even though I'm not playing, but yes, that's an eventuality. My stance would be that it's not worth trying to protect ourselves from this, but I might be under-considering the likeliness of such event or how bad it would be.

Let's say it's a one in a trillion occurrence, what's the consequence ? The parent connection might be closed by the child GC?

Reply:
The narrowness of the window assumes no-one else has installed an `after_fork`, or indeed done their own `fork` wrapping..

The consequence may be a segfault: the C drivers tend to Really Not Like It when someone external messes with the state of their internals, even if that someone is a forked copy of themselves. And if not, it's still a broken (not closed, according to the process's local state) connection, which will at best mean an exception in something undeserving of one.

Given that it's just a one-liner to completely close the window, I think I'd prefer to keep the `check!`. For all but that one in a trillion, it's a single integer comparison; if it does do any work, we're already in "impossibly rare" territory, and better that causes wasteful computation / extra Redis connections than cross-process errors.

Reply:
> extra Redis connections

Well, I think @rafaelfranca concerns about this was that it might be some stuffs that aren't safe to do in a finalizer.

But OK, I get your arguments, I'll add the checks back in a couple hours.

Reply:
@matthewd should be good now!

Reply:
> some stuffs that aren't safe to do in a finalizer

Yeah.. I think I share that concern, but feel it's the lesser risk (partly just because it would cause a problem in _this_ process, instead of a different one). And this is still only relevant if you hit the fork-then-GC race.

If we do encounter problems, we can probably shift the work from the finalizer into a thread or something, but I'm happy to defer that until we see an issue, because they'll manifest in a way that blames this code, rather than an ambiguous bug in connection handling in general. :+1: